### ISSUE TITLE
Suggestions to make us change testing SSL/Ruby versions easily

### ISSUE THREAD

**Issue by junaruga** [ruby/openssl#224]
I have some suggestions to make us change testing SSL/Ruby versions easily.

### 1. Making us change ruby-openssl-docker's `Dockerfile` easily 

I sent a pull-request to improve the situation. https://github.com/ruby/openssl/pull/223

### 2. Reducing the build time of `docker build` for ruby-openssl-docker

Right now each installation logic in the `Dockerfile` is sequential.

We can run each installation in parallel with Gnu Parallel. This can reduce the build time.

According to the tutorial document [1], the each `echo` command is run like this.

```
$ parallel echo ::: 2.3.7 2.4.4 2.5.1
2.3.7
2.4.4
2.5.1
```

So, for example below kind of logic in the `Dockerfile` is possible.

tool/ruby-openssl-docker/Dockerfile

```
ENV OPENSSL10_VERSIONS 1.0.0t 1.0.1u 1.0.2o
ENV OPENSSL11_VERSIONS 1.1.0h 1.1.1-pre8
ENV LIBRESSL_VERSIONS 2.3.10 2.4.5 2.5.5 2.6.5 2.7.4
ENV RUBY_VERSIONS 2.3.7 2.4.4 2.5.1

RUN parallel ./install_openssl10.sh ::: ${OPENSSL10_VERSIONS}
RUN parallel ./install_openssl11.sh ::: ${OPENSSL11_VERSIONS}
RUN parallel ./install_libressl.sh ::: ${LIBRESSL_VERSIONS}
RUN parallel ./install_ruby.sh ::: ${RUBY_VERSIONS}
```

### 3. Build and push new docker image automatically conditionally on Travis CI

Right now maybe we have to build and push the docker image to `zzak/ruby-openssl-docker:testing` manually. This is a little hard.

According to a project's case [2], I think that we can build the docker image in `after_success` process conditionally.
For example in `after_success` process, if the testing SSL/Ruby versions are changed in `.travis.yml`, the `docker build` command is run with the new versions, and pushed. [3] is the case with `docker build` with arguments.

How do you think?

## References

* [1] https://www.gnu.org/software/parallel/parallel_tutorial.html#Input-sources
* [2] https://github.com/COMBINE-lab/salmon/blob/master/.travis.yml#L45-L46
* [3] https://github.com/sclorg/rpm-list-builder/blob/master/.travis.yml#L31-L36

**Comment by rhenium** [comment]
> 1. Making us change ruby-openssl-docker's `Dockerfile` easily

Thank you. It looks nice.

> 2. Reducing the build time of `docker build` for ruby-openssl-docker

I don't know how to use GNU Parallel. Is it possible to have the build script inline (not having separate install_*.sh scripts)? Or, can we perhaps simply give `-jX` option to `make`?

Building the Docker image currently takes much time and a method that would speed up that is welcome. :)

> 3. Build and push new docker image automatically conditionally on Travis CI

I don't think this is a good idea. In this way, a new Docker image is built when our test suite passes using the old Docker image pulled from Docker Hub. This doesn't actually test the new image, and if it is broken, (irrelevant) subsequent CI builds start to fail. Also, I wonder what would happen if one restarted a CI build that contains old configuration in .travis.yml.

**Comment by junaruga** [comment]
> I don't know how to use GNU Parallel. Is it possible to have the build script inline (not having separate install_*.sh scripts)? Or, can we perhaps simply give -jX option to make?

I am also at first time to use the Gnu Parallel. :)
It might be possible to use it with the build script inline, seeing [1]. (And it looks better than separated install_*.sh scripts in this case). Also `make -jX` might be better than that.
I will want to do the experiment for the cases, then report here.

> I don't think this is a good idea. In this way, a new Docker image is built when our test suite passes using the old Docker image pulled from Docker Hub. This doesn't actually test the new image, and if it is broken, (irrelevant) subsequent CI builds start to fail. Also, I wonder what would happen if one restarted a CI build that contains old configuration in .travis.yml.

Ah I see. Your concern makes sense. I just had an idea that I might be able to automate to create the docker image automatically. But it might not be possible.

* [1] https://en.wikipedia.org/wiki/GNU_parallel

**Comment by junaruga** [comment]
> 2. Reducing the build time of docker build for ruby-openssl-docker

I did the experiment for below 2 cases with below commands.
The 2 cases are modified in addition, after above 1's pull-request.

```
$ nproc
4
$ cd tool/ruby-openssl-docker
$ time docker build --rm --no-cache -t ruby-openssl-docker .
```

Here is the result.

* Original: 36 minutes
* with `make` -jX`: 27 minutes ( the modification: https://github.com/junaruga/openssl/commit/5a79962 on feature/docker-build-time-1 branch )
* with `parallel` command: 21 minutes ( the modification:  https://github.com/junaruga/openssl/commit/b6ffb5b on feature/docker-build-time-2 branch )

How do you think?

**Comment by junaruga** [comment]
> 3. Build and push new docker image automatically conditionally on Travis CI

I got an idea to automate the processes to build and push `ruby-openssl-docker` docker image.
It's might be a full automation as a possibility for the processes.

## Step 1: Prepare

In a pull-request, if below files are included in the modification (= `BUILD_RUBY_OPENSSL_DOCKER` is `true`), build ruby-openssl-docker image from `tool/ruby-openssl-docker/Dockerfile`.

```
tool/ruby-openssl-docker/Dockerfile
tool/ruby-openssl-docker/init.sh
```
Then push the image to a DockerHub as a name: `$DOCKER_USER/ruby-openssl-docker:pr`.

## Step 2: Test

If `BUILD_RUBY_OPENSSL_DOCKER` is `true`, use testing docker image: `$DOCKER_USER/ruby-openssl-docker:pr` instead of instead of `$DOCKER_USER/ruby-openssl-docker:testing`.

If `BUILD_RUBY_OPENSSL_DOCKER` is not `true`, or DOCKER_USER and DOCKER_PASSWORD are not set, the behavior is same with current one.

## Step 3: Deploy

When merging a pull-request from forked repository, Travis is run as `push` event.
If `BUILD_RUBY_OPENSSL_DOCKER` is `true`, rename `$DOCKER_USER/ruby-openssl-docker:pr` to `$DOCKER_USER/ruby-openssl-docker:testing`.
But we carefully need to set this job's condition (`if` element) to prevent unintended run.

Seeing the document https://docs.travis-ci.com/user/conditions-v1
For example, we might be able to restrict executable users for this job.

We can run below kind of script without this job as semi automation, as this job can be run

```
docker pull $DOCKER_USER/ruby-openssl-docker:pr
docker tag $IMAGE:pr $IMAGE:testing
docker push $DOCKER_USER/ruby-openssl-docker:testing
```


## Demo

Here is the demo I prepared for my repository.

* `.travis.yml`: https://github.com/junaruga/openssl/blob/feature/docker-automate-building-demo/.travis.yml
* pull-request: You can see 3 commits. This is to add 2 commits of another PR: ( https://github.com/ruby/openssl/pull/223 ): https://github.com/junaruga/openssl/pull/2/commits
* Travis: You can see new OpenSSL 1.1.1 is actually tested in Travis page. https://travis-ci.org/junaruga/openssl/builds/452879550
* DockerHub: You can see both `pr` and `testing` tag. https://hub.docker.com/r/junaruga/ruby-openssl-docker/tags/

You need to `DOCKER_USER` and `DOCKER_PASSWORD` on Travis setting page like this.
This is the id and password of your DockerHub account.
If those are not set, the behavior is same with current one.

![travis-env-docker-user-password](https://user-images.githubusercontent.com/121989/48270955-8592e480-e43b-11e8-9f36-bfedc0e272fb.png)

**Comment by rhenium** [comment]
Ultimately we migrated from Travis CI + docker-compose to GitHub Actions, where we take Ruby binary from https://github.com/ruby/setup-ruby and build OpenSSL/LibreSSL on the fly.
